compiler/modules/print.llvm/llvm.bal

```diff
@@ -71,9 +71,53 @@ public function constInt(IntType ty, int val) returns ConstValue {
 
 // Corresponds to LLVMConstReal
 public function constReal(RealType ty, float val) returns ConstValue {
-    // XXX #307 use decimal for non-special floats 
-    return new ConstValue(ty, "0x" + int:toHexString(float:toBitsInt(val)));
-
+    string valRep;
+    if val.isInfinite() {
+        if val > 0.0 {
+            valRep = "0x7FF0000000000000";
+        }
+        else {
+            valRep = "0xFFF0000000000000";
+        }
+    }
+    else if val.isNaN() {
+        valRep = "0x7FF8000000000000";
+    }
+    else {
+        float exp = float:floor(float:log10(float:abs(val)));
```

Comment:
This code is far too difficult to understand. I have no idea what is doing.  It either needs a lot of comments or, probably better, do it in a simpler way. I don't understand why it needs to be this complicated.  Can you not convert the float to a string and then based on the string decide whether to use that or the hexadecimal representation?

Reply:
Moved the part which decide whether to hex to not to the top and added comments to clarify what each thing is doing. Most of it is to convert the number to scientific notation in the format we need if we are not using the hex representation.